src/Planner/CollectSets.cpp

```diff
@@ -51,11 +51,16 @@ class CollectSetsVisitor : public ConstInDepthQueryTreeVisitor<CollectSetsVisito
         auto * second_argument_table = in_second_argument->as<TableNode>();
         StorageSet * storage_set = second_argument_table != nullptr ? dynamic_cast<StorageSet *>(second_argument_table->getStorage().get()) : nullptr;
 
+        auto set_key = in_second_argument->getTreeHash();
+        /// For some cases we will get prepared set by AST Hash.
+        /// For example: queries disabling query_plan_optimize_primary_key or queries on StorageSystemNumbers.
+        auto ast_set_key = in_second_argument->toAST({.fully_qualified_identifiers = false})->getTreeHash();
```

Comment:
Why do you need it? This looks strange, I suppose you fail to find the set somewhere?

Reply:
@novikd Thanks for your review. 

Prepared sets will be used in `KeyCondition` to prune mark ranges. For new analyzer it use `QueryTreeNode` as key to get it  which I found 2 bad cases:

1. In this PR
I use `KeyCondition` to prune number range. And `KeyCondition` is constructing by `QuerySelectInfo` which means it is based on AST. When `KeyCondition` try to build RPN it will try to get prepared sets by AST tree hash.

2. Queries with settings query_plan_optimize_primary_key=false and allow_experimental_analyzer=true
The following queries will fail to prune mark.
```
SELECT woo.distinct_in_order.a AS a FROM woo.distinct_in_order WHERE woo.distinct_in_order.a IN (1) LIMIT _CAST(1, 'UInt64') SETTINGS query_plan_optimize_primary_key = 0
```

```
2023.08.22 15:02:07.480624 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Trace> Planner: Query SELECT woo.distinct_in_order.a AS a FROM woo.distinct_in_order WHERE woo.distinct_in_order.a IN (1) LIMIT _CAST(1, 'UInt64') SETTINGS query_plan_optimize_primary_key = 0, allow_experimental_analyzer = 1 from stage FetchColumns to stage Complete
2023.08.22 15:02:07.486522 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Trace> QueryPlanOptimizePrewhere: The min valid primary key position for moving to the tail of PREWHERE is 0
2023.08.22 15:02:07.490152 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Debug> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Key condition: unknown
2023.08.22 15:02:07.491445 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Debug> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Selected 2/2 parts by partition key, 2 parts by primary key, 8/8 marks by primary key, 8 marks to read from 2 ranges
2023.08.22 15:02:07.491702 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Spreading mark ranges among streams (default reading)
2023.08.22 15:02:07.492481 [ 1893547 ] {7ea6d27a-36a6-47d4-a408-44a0535b24a5} <Debug> MergeTreeReadPool: min_marks_for_concurrent_read=16384
```

After the commit, mark pruning works:

```
2023.08.22 15:18:46.045166 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> Planner: Query SELECT woo.distinct_in_order.a AS a FROM woo.distinct_in_order WHERE woo.distinct_in_order.a IN (1) LIMIT _CAST(1, 'UInt64') SETTINGS query_plan_optimize_primary_key = 0, allow_experimental_analyzer = 1 to stage Complete
2023.08.22 15:18:46.045796 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> ContextAccess (default): Access granted: SELECT(a) ON woo.distinct_in_order
2023.08.22 15:18:46.046679 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> Planner: Query SELECT woo.distinct_in_order.a AS a FROM woo.distinct_in_order WHERE woo.distinct_in_order.a IN (1) LIMIT _CAST(1, 'UInt64') SETTINGS query_plan_optimize_primary_key = 0, allow_experimental_analyzer = 1 from stage FetchColumns to stage Complete
2023.08.22 15:18:46.049635 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> QueryPlanOptimizePrewhere: The min valid primary key position for moving to the tail of PREWHERE is 0
2023.08.22 15:18:46.051823 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Debug> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Key condition: (column 0 in 1-element set)
2023.08.22 15:18:46.052354 [ 2092457 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Running binary search on index range for part all_9_9_0 (8 marks)
2023.08.22 15:18:46.052354 [ 2092465 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Running binary search on index range for part all_10_10_0 (2 marks)
2023.08.22 15:18:46.052574 [ 2092465 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found (LEFT) boundary mark: 0
2023.08.22 15:18:46.052645 [ 2092457 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found (LEFT) boundary mark: 0
2023.08.22 15:18:46.052701 [ 2092465 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found (RIGHT) boundary mark: 1
2023.08.22 15:18:46.052829 [ 2092465 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found continuous range in 2 steps
2023.08.22 15:18:46.052838 [ 2092457 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found (RIGHT) boundary mark: 2
2023.08.22 15:18:46.052977 [ 2092457 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Found continuous range in 6 steps
2023.08.22 15:18:46.053347 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Debug> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Selected 2/2 parts by partition key, 2 parts by primary key, 3/8 marks by primary key, 3 marks to read from 2 ranges
2023.08.22 15:18:46.053488 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Trace> woo.distinct_in_order (1a32111f-e69a-4ebf-b759-6c9edc719797) (SelectExecutor): Spreading mark ranges among streams (default reading)
2023.08.22 15:18:46.053914 [ 2091508 ] {760cdbd5-6160-4d55-b845-1ba59f5b1610} <Debug> MergeTreeReadPool: min_marks_for_concurrent_read=16384
```